text/0000-panic-plan.md

```diff
@@ -0,0 +1,187 @@
+This also means that the panic of an assert can be accidentally 'hijacked' by a locally defined `panic!` macro.
+
+*💔 **Problem 7:** `assert!` and related macros need to choose between `core::panic!` and `std::panic!`, and can't use `$crate::panic!` for proper hygiene.*
+
+## Implicit formatting arguments
+
+[RFC 2795] adds implicit formatting args, as follows:
+
+```rust
+let a = 4;
+println!("a is {a}");
+```
+
+It modifies `format_args!()` to automatically capture variables that are named in a formatting placeholder.
+
+With the current implementations of `panic!()` (both core's and std's), this would not work if there are no additional explicit arguments:
+
+```rust
+let a = 4;
+
+println!("{}", a); // prints `4`
+panic!("{}", a); // panics with `4`
+
+println!("{a}"); // prints `4`
+panic!("{a}"); // panics with `{a}`  😕
+
+println!("{a} {}", 4); // prints `4 4`
+panic!("{a} {}", 4); // panics with `4 4`
+```
+
+*💔 **Problem 8:** `panic!("error: {error}")` will silently not work as expected, after [RFC 2795] is implemented.*
+
+## Bloat
+
+`core::panic!("hello {")` produces the same `fmt::Arguments` as `format_args!("hello {{")`, not `format_args!("{}", "hello {")` to avoid pulling in string's `Display` code,
+which can be quite big.
+
+However, `core::panic!(non_static_str)` does need to expand to `format_args!("{}", non_static_str)`, because `fmt::Arguments` requires a `'static` lifetime
+for the non-formatted pieces. Because the `panic!` `macro_rules` macro can't distinguish between non-`'static` and `'static` values,
+this optimization is only applied to what macro_rules consider a `$_:literal`, which does not include `concat!(..)` or `CONST_STR`.
+
+*💔 **Problem 9:** `const CONST_STR: &'static str = "hi"; core::panic!(CONST_STR)` works,
+but will silently result in a lot more generated code than `core::panic!("hi")`.
+(And also needs [special handling](https://github.com/rust-lang/rust/pull/78069) to make `const_panic` work.)*
+
+# Solution if we could go back in time
+
+None of these these problems would have existed if
+1\) `panic!()` did not handle the single-argument case differently, and
+2\) `std::panic!` was no different than `core::panic!`:
+
+```rust
+// core
+macro_rules! panic {
+    () => (
+        $crate::panic!("explicit panic")
+    );
+    ($($t:tt)*) => (
+        $crate::panicking::panic_fmt($crate::format_args!($($t)+))
+    );
+}
+
+// std
+use core::panic;
+```
+
+The examples from problems 1, 2, 3, 4, 5, 6 and 9 would simply not compile, and problems 7 and 8 would not occur.
+
+However, that would break too much existing code.
+
+# Proposed solution
+
+Considering we should not break existing code, I propose we gate the breaking changs on the 2021 edition.
+
+In addition, we add a lint that *warns* about the problems in Rust 2015/2018, while not giving errors or changing the behaviour.
+
+Specifically:
+
+- Only for Rust 2021, we apply the breaking changes as in the previous section.
```

Comment:
How will this interact with macro expansion? For example, if I have the following code:


```rust
// crate2018 (2018 edition):

macro_rules! custom_panic {
    ($($arg:tt),*) => {
        panic!($($arg),*)
    }
}

fn foo() {
	custom_panic!("Weird format: {}");
}


// crate2021 (2021 edition)
```
fn bar() {
	custom_panic!("Weird format: {}");
}
```

Reply:
That's a good question. We could either look at the edition of the span of `panic` or at the edition of the span of `"Weird format: {}"`.

I think the former would make most sense. Then the macro doesn't change its behaviour until it is updated to Rust 2021 itself.

What do you think?

Reply:
That sounds reasonable to me. Note that this will be a 'best effort' approach to preventing breakage. With a somewhat contrived macro, this approach can be made to fail:

```rust
macro_rules! custom_invoke {
    ($name:ident) => {
        $name!("My panic: {}")
    }
}

fn main() {
    custom_invoke!(panic)
}
```

That is, a higher-order macro would use the ident `panic` from a 2021 edition crate, causing the new behavior to be used.

While this particular macro seems very unrealistic, there are some places in rustc that actually use higher-order macros. Unfortunately, I think we need to accept some amount of breakage if this RFC is accepted.

Reply:
That'd still make sense right? You're now passing the macro the 2021 panic, not the 2018 panic. With this style of macro, it's the responsibility of the caller to provide it with the name of a macro that works. Just like `custom_invoke!(println)` also wouldn't compile. So I'd say the breakage in this example is in the 2021 crate that calls the macro, not in the 2018 one that defines it.

Reply:
With proc macros, it may not be as clear that `crate2021` is to blame. Using `Span::resolved_at`, you could get a `panic` ident that appears to come from a 2018 crate, even if the literal `panic` never appears in a 2018 crate.

To be clear, I woud be shocked if this ever came up in practice. However, I think it would be worth adding an note to the RFC that this is idended to mitigate cross-edition breakage, not prevent it entirely.

Reply:
@Aaron1011 isn't this true for *all* edition changes though? If I write a proc-macro that attaches the wrong spans, then it may stop working when a crate using that macro upgrades to a new edition.

Perhaps there should be an exception to the stability rules for bugs that cause code to be tagged with the wrong edition.

Reply:
That's true. However, there have been a bunch of proc-macro API stabilizations since the 2018 edition, so I don't think it's really come up before.